### ISSUE TITLE
[proc-macro] range end index 4 out of range for slice of length 3

### ISSUE THREAD

**Issue by eminence** [rust-analyzer/rust-analyzer#10982]
I don't have a standalone reproducer yet, but here's something small using the `durable_worker` proc macro from the [worker](https://crates.io/crates/worker) crate:

src/main.rs:

```rust
use worker::*;

#[durable_object]
pub struct Counter {
}
 #[durable_object]
impl DurableObject for Counter {
    fn new(state: State, env: Env) -> Self {
        todo!()
    }

    async fn fetch(&mut self, _req: Request) -> Result<Response> {
        todo!()
    }
}

fn main() {
    println!("Hello, world!");
}
```

Cargo.toml:
```toml
[dependencies]
worker = "0.0.7"
```

Running `rust-analyzer diagnostics .` products many pages of trackback, but the bottom says:

```
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
processing crate: foo, module: /home/achin/tmp/17/foo/src/main.rs
Diagnostic { code: DiagnosticCode("macro-error"), message: "range end index 4 out of range for slice of length 3", range: 16..56, severity: Error, unused: false, experimental: true, fixes: None }
Diagnostic { code: DiagnosticCode("macro-error"), message: "internal error: entered unreachable code", range: 58..267, severity: Error, unused: false, experimental: true, fixes: None }

diagnostic scan complete

[ERROR rust_analyzer] Unexpected error: diagnostic error detected
diagnostic error detected
```

If you want to see the full output, let me know.

**rust-analyzer version**: 021d5821befc2e0a0d7d164b345f93d8dfaac52a

**rustc version**: rustc 1.58.0-nightly (936238a92 2021-11-11)

**Comment by bjorn3** [comment]
Try updating to the latest nightly rustc. There was a change to the proc macro abi a little less than a month ago which rust-analyzer has already updated to. Rust-analyzer at any point in time only supports a single nightly version in addition to several stable (and latest beta) versions of rustc.

**Comment by eminence** [comment]
Shoot, I keep forgetting RA depends on nightly rustc.    That fixed it, sorry for the noise.

**Comment by flodiebold** [comment]
It doesn't, we also supports stable, but if you _do_ use nightly, it needs to be up to date.
